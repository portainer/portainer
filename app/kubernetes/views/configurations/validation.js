import { KubernetesSecretTypeOptions } from '@/kubernetes/models/configuration/models';
import { KubernetesConfigurationKinds } from '@/kubernetes/models/configuration/models';

export function isConfigurationFormValid(alreadyExist, isDataValid, formValues) {
  const uniqueCheck = !alreadyExist && isDataValid;
  let secretWarningMessage = '';
  let isFormValid = false;

  if (formValues.IsSimple) {
    if (formValues.Kind === KubernetesConfigurationKinds.SECRET) {
      let isSecretDataValid = true;

      switch (formValues.Type) {
        case KubernetesSecretTypeOptions.SERVICEACCOUNTTOKEN.value:
          // data isn't required for service account tokens
          isFormValid = uniqueCheck && formValues.ResourcePool;
          return [isFormValid, ''];
        case KubernetesSecretTypeOptions.DOCKERCFG.value:
          // needs to contain a .dockercfg key
          isSecretDataValid = formValues.Data.some((entry) => entry.Key === '.dockercfg');
          secretWarningMessage = isSecretDataValid ? '' : 'A data entry with a .dockercfg key is required.';
          break;
        case KubernetesSecretTypeOptions.DOCKERCONFIGJSON.value:
          // needs to contain a .dockerconfigjson key
          isSecretDataValid = formValues.Data.some((entry) => entry.Key === '.dockerconfigjson');
          secretWarningMessage = isSecretDataValid ? '' : 'A data entry with a .dockerconfigjson key. is required.';
          break;
        case KubernetesSecretTypeOptions.BASICAUTH.value:
          isSecretDataValid = formValues.Data.some((entry) => entry.Key === 'username' || entry.Key === 'password');
          secretWarningMessage = isSecretDataValid ? '' : 'A data entry with a username or password key is required.';
          break;
        case KubernetesSecretTypeOptions.SSHAUTH.value:
          isSecretDataValid = formValues.Data.some((entry) => entry.Key === 'ssh-privatekey');
          secretWarningMessage = isSecretDataValid ? '' : `A data entry with a 'ssh-privatekey' key is required.`;
          break;
        case KubernetesSecretTypeOptions.TLS.value:
          isSecretDataValid = formValues.Data.some((entry) => entry.Key === 'tls.crt') && formValues.Data.some((entry) => entry.Key === 'tls.key');
          secretWarningMessage = isSecretDataValid ? '' : `Data entries containing a 'tls.crt' key and a 'tls.key' key are required.`;
          break;
        case KubernetesSecretTypeOptions.BOOTSTRAPTOKEN.value:
          isSecretDataValid = formValues.Data.some((entry) => entry.Key === 'token-id') && formValues.Data.some((entry) => entry.Key === 'token-secret');
          secretWarningMessage = isSecretDataValid ? '' : `Data entries containing a 'token-id' key and a 'token-secret' key are required.`;
          break;
        default:
          break;
      }

      isFormValid = uniqueCheck && formValues.ResourcePool && formValues.Data.length >= 1 && isSecretDataValid;
      return [isFormValid, secretWarningMessage];
    }

    isFormValid = formValues.Data.length > 0 && uniqueCheck && formValues.ResourcePool;
    return [isFormValid, secretWarningMessage];
  }

  isFormValid = uniqueCheck && formValues.ResourcePool;
  return [isFormValid, secretWarningMessage];
}
