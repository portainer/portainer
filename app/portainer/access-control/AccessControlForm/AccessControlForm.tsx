import { FormikErrors } from 'formik';

import { useUser } from '@/portainer/hooks/useUser';
import { EditDetails } from '@/portainer/access-control/EditDetails/EditDetails';

import { FormSectionTitle } from '@@/form-components/FormSectionTitle';
import { SwitchField } from '@@/form-components/SwitchField';

import { ResourceControlOwnership, AccessControlFormData } from '../types';

export interface Props {
  values: AccessControlFormData;
  onChange(values: AccessControlFormData): void;
  hideTitle?: boolean;
  formNamespace?: string;
  errors?: FormikErrors<AccessControlFormData>;
}

export function AccessControlForm({
  values,
  onChange,
  hideTitle,
  formNamespace,
  errors,
}: Props) {
  const { isAdmin } = useUser();

  const accessControlEnabled =
    values.ownership !== ResourceControlOwnership.PUBLIC;
  return (
    <>
      {!hideTitle && <FormSectionTitle>Access control</FormSectionTitle>}

      <div className="form-group">
        <div className="col-sm-12">
          <SwitchField
            checked={accessControlEnabled}
            name={withNamespace('accessControlEnabled')}
            label="Enable access control"
            tooltip="When enabled, you can restrict the access and management of this resource."
            onChange={handleToggleEnable}
            dataCy="portainer-accessMgmtToggle"
          />
        </div>
      </div>

      {accessControlEnabled && (
        <EditDetails
          onChange={onChange}
          values={values}
          errors={errors}
          formNamespace={formNamespace}
        />
      )}
    </>
  );

  function withNamespace(name: string) {
    return formNamespace ? `${formNamespace}.${name}` : name;
  }

  function handleToggleEnable(accessControlEnabled: boolean) {
    let ownership = ResourceControlOwnership.PUBLIC;
    if (accessControlEnabled) {
      ownership = isAdmin
        ? ResourceControlOwnership.ADMINISTRATORS
        : ResourceControlOwnership.PRIVATE;
    }
    onChange({ ...values, ownership });
  }
}
