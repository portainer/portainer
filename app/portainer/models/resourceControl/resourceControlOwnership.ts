export enum ResourceControlOwnership {
  PUBLIC = 'public',
  PRIVATE = 'private',
  RESTRICTED = 'restricted',
  ADMINISTRATORS = 'administrators',
}
