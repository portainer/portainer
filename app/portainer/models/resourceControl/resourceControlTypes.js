export const ResourceControlTypeString = Object.freeze({
  CONFIG: 'config',
  CONTAINER: 'container',
  NETWORK: 'network',
  SECRET: 'secret',
  SERVICE: 'service',
  STACK: 'stack',
  VOLUME: 'volume',
  CUSTOM_TEMPLATE: 'custom-template',
  CONTAINER_GROUP: 'container-group',
});

/**
 * ResourceType int defined in portainer.go as ResourceControlType
 */
export const ResourceControlTypeInt = Object.freeze({
  CONTAINER: 1,
  SERVICE: 2,
  VOLUME: 3,
  NETWORK: 4,
  SECRET: 5,
  STACK: 6,
  CONFIG: 7,
  CUSTOM_TEMPLATE: 8,
  CONTAINER_GROUP: 9,
});
