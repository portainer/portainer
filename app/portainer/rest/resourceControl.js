angular.module('portainer.app').factory('ResourceControl', [
  '$resource',
  'API_ENDPOINT_RESOURCE_CONTROLS',
  function ResourceControlFactory($resource, API_ENDPOINT_RESOURCE_CONTROLS) {
    'use strict';
    return $resource(
      API_ENDPOINT_RESOURCE_CONTROLS + '/:id',
      {},
      {
        update: { method: 'PUT', params: { id: '@id' } },
      }
    );
  },
]);
