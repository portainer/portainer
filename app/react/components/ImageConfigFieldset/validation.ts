import { bool, number, object, SchemaOf, string } from 'yup';

import { Values } from './types';

export function validation(): SchemaOf<Values> {
  return object({
    image: string().required('Image is required'),
    registryId: number().default(0),
    useRegistry: bool().default(false),
  });
}
