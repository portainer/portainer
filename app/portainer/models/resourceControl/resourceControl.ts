import { ResourceControlOwnership as RCO } from './resourceControlOwnership';

export enum ResourceControlType {
  // Container represents a resource control associated to a Docker container
  Container = 1,
  // Service represents a resource control associated to a Docker service
  Service,
  // Volume represents a resource control associated to a Docker volume
  Volume,
  // Network represents a resource control associated to a Docker network
  Network,
  // Secret represents a resource control associated to a Docker secret
  Secret,
  // Stack represents a resource control associated to a stack composed of Docker services
  Stack,
  // Config represents a resource control associated to a Docker config
  Config,
  // CustomTemplate represents a resource control associated to a custom template
  CustomTemplate,
  // ContainerGroup represents a resource control associated to an Azure container group
  ContainerGroup,
}

export interface ResourceControlResponse {
  Id: number;
  Type: ResourceControlType;
  ResourceId: string | number;
  UserAccesses: unknown[];
  TeamAccesses: unknown[];
  Public: boolean;
  AdministratorsOnly: boolean;
  System: boolean;
}

export class ResourceControlViewModel {
  Id: number;

  Type: ResourceControlType;

  ResourceId: string | number;

  UserAccesses: unknown[];

  TeamAccesses: unknown[];

  Public: boolean;

  System: boolean;

  Ownership: RCO;

  constructor(data: ResourceControlResponse) {
    this.Id = data.Id;
    this.Type = data.Type;
    this.ResourceId = data.ResourceId;
    this.UserAccesses = data.UserAccesses;
    this.TeamAccesses = data.TeamAccesses;
    this.Public = data.Public;
    this.System = data.System;
    this.Ownership = determineOwnership(this);
  }
}

function determineOwnership(resourceControl: ResourceControlViewModel) {
  if (resourceControl.Public) {
    return RCO.PUBLIC;
  }

  if (
    resourceControl.UserAccesses.length === 1 &&
    resourceControl.TeamAccesses.length === 0
  ) {
    return RCO.PRIVATE;
  }

  if (
    resourceControl.UserAccesses.length > 1 ||
    resourceControl.TeamAccesses.length > 0
  ) {
    return RCO.RESTRICTED;
  }

  return RCO.ADMINISTRATORS;
}
