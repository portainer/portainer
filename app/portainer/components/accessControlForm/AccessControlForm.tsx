import _ from 'lodash';
import { useEffect, useState, useCallback } from 'react';

import { ResourceControlOwnership as RCO } from '@/portainer/models/resourceControl/resourceControlOwnership';
import { BoxSelector, buildOption } from '@/portainer/components/BoxSelector';
import { ownershipIcon } from '@/portainer/filters/filters';
import { useUser } from '@/portainer/hooks/useUser';
import { Team } from '@/portainer/teams/types';
import { UserViewModel } from '@/portainer/models/user';
import { BoxSelectorOption } from '@/portainer/components/BoxSelector/types';
import { FormSectionTitle } from '@/portainer/components/form-components/FormSectionTitle';
import { SwitchField } from '@/portainer/components/form-components/SwitchField';

import { AccessControlFormData } from './model';
import { UsersField } from './UsersField';
import { TeamsField } from './TeamsField';

export interface BaseProps {
  values: AccessControlFormData;
  onChange(values: AccessControlFormData): void;
  hideTitle?: boolean;
}

interface Props extends BaseProps {
  teams: Team[];
  users: UserViewModel[];
}

export function AccessControlForm({
  values,
  onChange,
  hideTitle,
  users,
  teams,
}: Props) {
  const { user } = useUser();
  const isAdmin = user?.Role === 1;

  const options = useOptions(isAdmin, teams);

  const handleChange = useCallback(
    (partialValues: Partial<typeof values>) => {
      onChange({ ...values, ...partialValues });
    },

    [values, onChange]
  );

  return (
    <>
      {!hideTitle && <FormSectionTitle>Access control</FormSectionTitle>}

      <div className="form-group">
        <div className="col-sm-12">
          <SwitchField
            checked={values.accessControlEnabled}
            name="ownership"
            label="Enable access control"
            tooltip="When enabled, you can restrict the access and management of this resource."
            onChange={(accessControlEnabled) =>
              handleChange({ accessControlEnabled })
            }
          />
        </div>
      </div>

      {values.accessControlEnabled && (
        <>
          <div className="form-group">
            <BoxSelector
              radioName="access-control"
              value={values.ownership}
              options={options}
              onChange={(ownership) => handleChange({ ownership })}
            />
          </div>
          {values.ownership === RCO.RESTRICTED && (
            <div aria-label="extra-options">
              {isAdmin && (
                <UsersField
                  users={users}
                  onChange={(authorizedUsers) =>
                    handleChange({ authorizedUsers })
                  }
                  value={values.authorizedUsers}
                />
              )}

              {(isAdmin || teams.length > 1) && (
                <TeamsField
                  teams={teams}
                  overrideTooltip={
                    !isAdmin && teams.length > 1
                      ? 'As you are a member of multiple teams, you can select which teams(s) will be able to manage this resource.'
                      : undefined
                  }
                  onChange={(authorizedTeams) =>
                    handleChange({ authorizedTeams })
                  }
                  value={values.authorizedTeams}
                />
              )}
            </div>
          )}
        </>
      )}
    </>
  );
}

function useOptions(isAdmin: boolean, teams: Team[]) {
  const [options, setOptions] = useState<Array<BoxSelectorOption<RCO>>>([]);

  useEffect(() => {
    setOptions(isAdmin ? adminOptions() : nonAdminOptions(teams));
  }, [isAdmin, teams]);

  return options;
}

function adminOptions() {
  return [
    buildOption(
      'access_administrators',
      ownershipIcon('administrators'),
      'Administrators',
      'I want to restrict the management of this resource to administrators only',
      RCO.ADMINISTRATORS
    ),
    buildOption(
      'access_restricted',
      ownershipIcon('restricted'),
      'Restricted',
      'I want to restrict the management of this resource to a set of users and/or teams',
      RCO.RESTRICTED
    ),
  ];
}
function nonAdminOptions(teams: Team[]) {
  return _.compact([
    buildOption(
      'access_private',
      ownershipIcon('private'),
      'Private',
      'I want to this resource to be manageable by myself only',
      RCO.PRIVATE
    ),
    teams.length > 0 &&
      buildOption(
        'access_restricted',
        ownershipIcon('restricted'),
        'Restricted',
        teams.length === 1
          ? `I want any member of my team (${teams[0].Name})  to be able to manage this resource`
          : 'I want to restrict the management of this resource to one or more of my teams',
        RCO.RESTRICTED
      ),
  ]);
}
