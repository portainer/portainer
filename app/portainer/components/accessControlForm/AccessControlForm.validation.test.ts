import { ResourceControlOwnership } from '@/portainer/models/resourceControl/resourceControlOwnership';

import { validationSchema } from './AccessControlForm.validation';

test('when access control is disabled, should be valid', async () => {
  const schema = validationSchema(true);
  const object = { accessControlEnabled: false };

  await expect(
    schema.validate(object, { strict: true })
  ).resolves.toStrictEqual(object);
});

test('when only access control is enabled, should be invalid', async () => {
  const schema = validationSchema(true);

  await expect(
    schema.validate({ accessControlEnabled: true }, { strict: true })
  ).rejects.toThrowErrorMatchingSnapshot();
});

test('when access control is enabled and ownership not restricted, should be valid', async () => {
  const schema = validationSchema(true);
  [
    ResourceControlOwnership.ADMINISTRATORS,
    ResourceControlOwnership.PRIVATE,
    ResourceControlOwnership.PUBLIC,
  ].forEach(async (ownership) => {
    const object = { accessControlEnabled: false, ownership };

    await expect(
      schema.validate(object, { strict: true })
    ).resolves.toStrictEqual(object);
  });
});

test('when access control is enabled, ownership is restricted and no teams or users, should be invalid', async () => {
  [true, false].forEach(async (isAdmin) => {
    const schema = validationSchema(isAdmin);

    await expect(
      schema.validate(
        {
          accessControlEnabled: true,
          ownership: ResourceControlOwnership.RESTRICTED,
          authorizedTeams: [],
          authorizedUsers: [],
        },
        { strict: true }
      )
    ).rejects.toThrowErrorMatchingSnapshot();
  });
});

test('when access control is enabled, ownership is restricted, user is admin should have either teams or users', async () => {
  const schema = validationSchema(true);
  const teams = {
    accessControlEnabled: true,
    ownership: ResourceControlOwnership.RESTRICTED,
    authorizedTeams: [1],
    authorizedUsers: [],
  };

  await expect(schema.validate(teams, { strict: true })).resolves.toStrictEqual(
    teams
  );

  const users = {
    accessControlEnabled: true,
    ownership: ResourceControlOwnership.RESTRICTED,
    authorizedTeams: [],
    authorizedUsers: [1],
  };

  await expect(schema.validate(users, { strict: true })).resolves.toStrictEqual(
    users
  );

  const both = {
    accessControlEnabled: true,
    ownership: ResourceControlOwnership.RESTRICTED,
    authorizedTeams: [1],
    authorizedUsers: [2],
  };

  await expect(schema.validate(both, { strict: true })).resolves.toStrictEqual(
    both
  );
});

test('when access control is enabled, ownership is restricted, user is admin with teams and users, should be valid', async () => {
  const schema = validationSchema(false);

  const object = {
    accessControlEnabled: true,
    ownership: ResourceControlOwnership.RESTRICTED,
    authorizedTeams: [1],
    authorizedUsers: [1],
  };

  await expect(
    schema.validate(object, { strict: true })
  ).resolves.toStrictEqual(object);
});

test('when access control is enabled, ownership is restricted, user is not admin with teams, should be valid', async () => {
  const schema = validationSchema(false);

  const object = {
    accessControlEnabled: true,
    ownership: ResourceControlOwnership.RESTRICTED,
    authorizedTeams: [1],
  };

  await expect(
    schema.validate(object, { strict: true })
  ).resolves.toStrictEqual(object);
});
