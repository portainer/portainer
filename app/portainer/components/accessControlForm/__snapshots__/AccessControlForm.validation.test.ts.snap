// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`when access control is enabled, ownership is restricted and no teams or users, should be invalid 1`] = `"You must specify at least one team"`;

exports[`when access control is enabled, ownership is restricted and no teams or users, should be invalid 2`] = `"You must specify at least one team"`;

exports[`when access control is enabled, ownership is restricted and no teams or users, should be valid 1`] = `"You must specify at least one team"`;

exports[`when access control is enabled, ownership is restricted and no teams or users, should be valid 2`] = `"You must specify at least one team"`;

exports[`when access control is enabled, ownership is restricted, user is admin but no users, should be valid 1`] = `"You must specify at least one user."`;

exports[`when only access control is enabled, should be invalid 1`] = `"ownership is a required field"`;
